src/Blockcore.Indexer.Core/Storage/Mongo/MongoStorageOperations.cs

```diff
@@ -126,60 +135,91 @@ public SyncBlockInfo PushStorageBatch(StorageBatch storageBatch)
+         }
+
+         Task transactionTableTask = Task.CompletedTask;
+
+         try
          {
-            try
+            if (storageBatch.TransactionTable.Any())
             {
-               if (storageBatch.TransactionTable.Any())
-                  data.TransactionTable.InsertMany(storageBatch.TransactionTable, new InsertManyOptions {IsOrdered = false});
+               transactionTableTask = data.TransactionTable.InsertManyAsync(storageBatch.TransactionTable,
+                  new InsertManyOptions { IsOrdered = false });
             }
-            catch (MongoBulkWriteException mbwex)
+         }
+         catch (MongoBulkWriteException mbwex)
+         {
+            // transactions are a special case they are not deleted from store in case of reorgs
+            // because they will just be included in another blocks, so we ignore if key is already present
+            if (mbwex.WriteErrors.Any(e => e.Category != ServerErrorCategory.DuplicateKey))
             {
-               // transactions are a special case they are not deleted from store in case of reorgs
-               // because they will just be included in another blocks, so we ignore if key is already present
-               if (mbwex.WriteErrors.Any(e => e.Category != ServerErrorCategory.DuplicateKey))
-               {
-                  throw;
-               }
+               throw;
             }
-         });
+         }
 
-         Task.WaitAll(t3, t4);
+         var utxos = new List<UnspentOutputTable>(storageBatch.OutputTable.Values.Count);
 
-         var t6 = Task.Run(() =>
+         foreach (OutputTable outputTable in storageBatch.OutputTable.Values)
          {
-            if (globalState.IndexModeCompleted)
+            utxos.Add(new UnspentOutputTable
             {
-               PipelineDefinition<InputTable, InputTable> pipeline = BlockIndexer.BuildInputsAddressUpdatePiepline();
-               data.InputTable.Aggregate(pipeline);
-            }
-         });
+               Address = outputTable.Address,
+               Outpoint = outputTable.Outpoint,
+               Value = outputTable.Value,
+               BlockIndex = outputTable.BlockIndex
+            });
+         }
 
-         Task.WaitAll(t1, t2, t3, t4, t5, t6);
+         var unspentOutputTableTask = utxos.Any()
+            ? data.UnspentOutputTable.InsertManyAsync(utxos)
+            : Task.CompletedTask;
+
+         Task.WaitAll(blockTableTask, transactionBlockTableTask, outputTableTask, inputTableTask, transactionTableTask, unspentOutputTableTask);
+
+         var outpointsFromNewInput = storageBatch.InputTable
+            .Select(_ => _.Outpoint)
+            .ToList();
+
+         if (outpointsFromNewInput.Any())
+         {
+            var filterToDelete = Builders<UnspentOutputTable>.Filter
+               .Where(_ => outpointsFromNewInput.Contains(_.Outpoint));
+
+            data.UnspentOutputTable.DeleteMany(filterToDelete);
```

Comment:
In theory you dont have to wait for all tasks to finish just `unspentOutputTableTask` before you can trigger this one right?

```
Task.WaitAll(unspentOutputTableTask)

data.UnspentOutputTable.DeleteMany..
```